docs/content/ingestion/locking-and-priority.md

```diff
@@ -24,30 +24,73 @@ title: "Task Locking & Priority"
 
 # Task Locking & Priority
 
+This document explains the task locking system in Druid. Druid's locking system
+and versioning system are tighly coupled with each other to guarantee the correctness of ingested data.
+
+## Overshadow Relation between Segments
+
+You can run a task to overwrite existing data. The segments created by an overwriting task _overshadows_ existing segments.
+Note that the overshadow relation holds only for the same time chunk and the same data source.
+These overshadowed segments are not considered in query processing to filter out stale data.
+
+A segment `s1` can overshadow another `s2` if
+
+- `s1` has a higher major version than `s2`.
+- `s1` has the same major version and a higher minor version than `s2`.
+
+Here are some examples.
+
+- A segment of the major version of `2019-01-01T00:00:00.000Z` and the minor version of `0` overshadows
+ another of the major version of `2018-01-01T00:00:00.000Z` and the minor version of `1`.
+- A segment of the major version of `2019-01-01T00:00:00.000Z` and the minor version of `1` overshadows
+ another of the major version of `2019-01-01T00:00:00.000Z` and the minor version of `0`.
+
 ## Locking
 
-Once an Overlord process accepts a task, the task acquires locks for the data source and intervals specified in the task.
+If you are running two or more [druid tasks](./tasks.html) which generate segments for the same data source and the same time chunk,
+the generated segments could potentially overshadow each other which could lead to incorrect query results.
 
-There are two lock types, i.e., _shared lock_ and _exclusive lock_.
+To avoid this problem, tasks should get locks first before creating any segment in Druid.
+There are two types of locks, i.e., _time chunk lock_ and _segment lock_, and each task can use different types of locks.
```

Comment:
> In the docs for the forceTimeChunkLock parameter, i see "If not set, each task automatically chooses a lock type to use", maybe that should be mentioned here?

Sounds good. Added some description about it.

> Otherwise I think we need a few more usage details here, and I have a few questions. Can any type of task use any type of lock? I think i remember it not being implemented for a handful of indexing task types in the original PR. But If yes, should any type of task use any type of lock? Should we offer advice on which sorts of locks are useful with which types of task if the answer to the 2nd question is "no"? If the answer to both of these questions is "yes" then this is fine.

 The answer is yes for both questions. Any type of task can use any type of lock. The segment locking is just not implemented for some task types yet.